composites/Plugin/SnippetEditor/tests/selectionTest.js

```diff
@@ -0,0 +1,210 @@
+
+	return text;
+}
+
+/**
+ * Tests the selectReplacementVariables function.
+ *
+ * @param {Object} args The arguments with which to test.
+ * @param {string} args.before   The editor state before the onChange,
+ *                               represented as text.
+ * @param {string} args.after    The editor state after/during the onChange,
+ *                               represented as text.
+ * @param {string} args.expected The expected state the function should return,
+ *                               represented as text.
+ * @returns {void}
+ */
+function expectToMatch( args ) {
+	const { before: beforeText, after: afterText, expected: expectedText, backwards = false } = args;
+
+	const text = "Text %%entity%% %%entity%% Text";
+	const editorState = unserializeEditor( text, [ { name: "entity", value: "EntityValue" } ] );
+	const key = editorState.getCurrentContent().getFirstBlock().getKey();
+
+	const before = convertToState( beforeText, key, editorState, backwards );
+	const after = convertToState( afterText, key, editorState, backwards );
+	const expected = convertToState( expectedText, key, editorState, backwards );
+
+	const actual = selectReplacementVariables( after, before );
+	const actualText = convertFromState( "Text entity entity Text", actual.getSelection() );
+
+	expect( actualText ).toEqual( expectedText );
+	expect( actual.getSelection() ).toEqual( expected.getSelection() );
+
+	if ( backwards === false ) {
+		expectToMatch( { ...args, backwards: true } );
+	}
+}
+
+describe( "selection behavior", () => {
+	it( "selects the whole entity when selecting a small part from the right", () => {
+		expectToMatch( {
+			before: "Text entity[] entity Text",
+			after: "Text entit[y] entity Text",
+			expected: "Text [entity] entity Text",
+		} );
+	} );
+
+	it( "selects the whole entity when selecting a small part from the left", () => {
+		expectToMatch( {
+			before: "Text []entity entity Text",
+			after: "Text [e]ntity entity Text",
+			expected: "Text [entity] entity Text",
+		} );
+	} );
+
+	it( "moves over an entity to the right", () => {
+		expectToMatch( {
+			before: "Text []entity entity Text",
+			after: "Text e[]ntity entity Text",
+			expected: "Text [entity] entity Text",
+		} );
+	} );
+
+	it( "moves over an entity to the left", () => {
+		expectToMatch( {
+			before: "Text entity[] entity Text",
+			after: "Text entit[]y entity Text",
+			expected: "Text [entity] entity Text",
+		} );
+	} );
+
+	it( "selects the whole entity when clicking it", () => {
+		expectToMatch( {
+			before: "[]Text entity entity Text",
+			after: "Text ent[]ity entity Text",
+			expected: "Text [entity] entity Text",
+		} );
+	} );
+
+	it( "selects multiple entities when selecting half-way", () => {
+		expectToMatch( {
+			before: "[]Text entity entity Text",
+			after: "Text ent[ity ent]ity Text",
+			expected: "Text [entity entity] Text",
+		} );
+	} );
+
+	it( "moves out of selection when pressing arrow keys", () => {
```

Comment:
Or clicking before/after it. We don't know whether it's done by the arrow keys or the mouse/